{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "d71e8f5f",
   "metadata": {},
   "source": [
    "# Pattern matching\n",
    "\n",
    "\n",
    "We will start in the same way as the last notebook started  - by downloading/importing the packages needed and importing the .csv file(s) needed. In this case, we only need the .csv file that has the matched abstracts as we are specifically looking at person-first and identity-first patterns that are \"about\" autism (or ASD, Asperger's syndrome, etc.). \n",
    "\n",
    "We could use the same basic approach to look at person-first and identity-first language for other conditions for which there are good noun and adjective forms of the words (diabetes? obesity? cancer? something else?). Doing that would mean using the .csv file with all of the abstracts or potentially creating and entirely new file of abstracts matched to another condition of interest. However, that lies outside the scope of this research, so I will not address it further here. \n",
    "\n",
    "## Get ready \n",
    "\n",
    "As always, we start with code that:\n",
    "* loads up and nicknames some useful packages, \n",
    "* checks file locations,\n",
    "* imports files, and \n",
    "* checks them. \n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "271c801d",
   "metadata": {},
   "outputs": [],
   "source": [
    "%%capture\n",
    "\n",
    "!pip install nltk\n",
    "!pip install spacy -q\n",
    "!python -m spacy download en_core_web_lg -q\n",
    "\n",
    "import os                         # os is a module for navigating your machine (e.g., file directories).\n",
    "import nltk                       # nltk stands for natural language tool kit and is useful for text-mining. \n",
    "from nltk import word_tokenize    # and some of its key functions\n",
    "from nltk import sent_tokenize  \n",
    "tokenizer = nltk.tokenize.punkt.PunktSentenceTokenizer()\n",
    "from nltk.tokenize import sent_tokenize\n",
    "nltk.download('punkt')\n",
    "from nltk.corpus import wordnet                    # Finally, things we need for lemmatising!\n",
    "from nltk.stem import WordNetLemmatizer\n",
    "lemmatizer = WordNetLemmatizer() \n",
    "from nltk.stem.porter import PorterStemmer\n",
    "porter = PorterStemmer()\n",
    "nltk.download('averaged_perceptron_tagger')        # Like a POS-tagger...\n",
    "nltk.download('wordnet')\n",
    "nltk.download('webtext')\n",
    "from nltk.corpus import webtext\n",
    "\n",
    "import pandas as pd\n",
    "pd.set_option('display.max_colwidth', 200)\n",
    "import numpy as np\n",
    "import statistics\n",
    "\n",
    "import csv                        # csv is for importing and working with csv files\n",
    "\n",
    "from collections import Counter\n",
    "\n",
    "import statistics\n",
    "import re                         # things we need for RegEx corrections\n",
    "import string \n",
    "import spacy \n",
    "from spacy.matcher import Matcher \n",
    "from spacy.tokens import Span \n",
    "from spacy import displacy \n",
    "nlp = spacy.load('en_core_web_lg')\n",
    "nlp.max_length = 1500000 #or any large value, as long as you don't run out of RAM\n",
    "\n",
    "import math\n",
    "import matplotlib.pyplot as plt\n",
    "print(os.listdir(\"..\\\\output\")  )    \n",
    "\n",
    "# the '%%capture' at the top of this cell suppresses the output (which is normally quite long and annoying looking). \n",
    "# You can remove or comment it out if you prefer to see the output."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8e4e4ea5",
   "metadata": {},
   "source": [
    "## Import\n",
    "\n",
    "Having checked the contents of the output folder and seen the files we expected to see, we can now import the specific file of interest for this step of the analysis."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "75b74ce9",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "3794"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "matched_texts = pd.read_csv('..\\\\output\\\\matched_abstracts_no_null_texts.csv')    # one for just those that match the keyword\n",
    "len(matched_texts)                                                                # check the length "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f5d89f25",
   "metadata": {},
   "source": [
    "## Cleaning phase\n",
    "\n",
    "Cleaning begins by turning any instances of extra whitespaces (two or more in a row) into a single whitespace. Then, identifying any run-on sentences (where a lowercase letter, a full stop, and an uppercase letter are clustered without a whitespace) and inserting a whitespace between the full stop and the uppercase letter. Both of these steps will improve the sentence tokenisation that happens next. \n",
    "\n",
    "Then, we proceed to sentence tokenising the text. Like word tokens, sentence tokens become the unit for analysisis. As a trivial example, sentence tokenisation would turn a short text such as \n",
    "\n",
    "\n",
    "''' The cat named Cat is one of five cats. Honestly, I wonder why I have so many cats.\n",
    "''' \n",
    "\n",
    "into a list of sentence tokens like\n",
    "\n",
    "''' [[The cat named Cat is one of five cats.]\n",
    "\n",
    "[Honestly, I wonder why I have so many cats.]]\n",
    "\n",
    "''' \n",
    "\n",
    "An important difference is that the punctuation within the sentences that contributes to its structured and meaning (e.g. the comma and the full stops) are retained. This punctuation, like the capitalisation at the start of the sentences or for the poper nouns, is also retained as it helps the sentence-tokenisation process identify the words within the sentence correctly for their parts of speech (e.g. which of the words are nouns, verbs, etc. ). \n",
    "\n",
    "\n",
    "\n",
    "The sentence tokens are then put on individual rows, filtered to retain only those that contain one or more of the keywords of interest, and then filtered to ensure that there are no empty rows or duplicates. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "d6cfa42e",
   "metadata": {},
   "outputs": [],
   "source": [
    "def remove_errors (input):\n",
    "    no_extra_spaces = re.sub(r'(\\s)(\\s+)', r'\\1', input)               # turn 2+ sequential whitespaces into 1\n",
    "    no_run_ons1 = re.sub(r'([a-z].)([A-Z])', r'\\1 \\2', no_extra_spaces) # identifies run-ons (e.g. \"word.New sentence \")\n",
    "    no_run_ons2 = re.sub(r'([A-Z].)([A-Z])', r'\\1. \\2', no_run_ons1) # identifies run-ons (e.g. \"ACRONYM.New sentence \")\n",
    "\n",
    "    return(no_run_ons2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "d2edc6ca",
   "metadata": {},
   "outputs": [],
   "source": [
    "no_run_ons = [remove_errors(abstract) for abstract in matched_texts['Text'] ] \n",
    "                                             # create abstract list without extra spaces/run-ons \n",
    "                                             # this is to improve sentence tokenisation later \n",
    "matched_texts['Sentence'] = no_run_ons       # copy the no extra space/run-on abstract list back into df as a new column"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "17421e4d",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "How many sentences in total: 56473\n"
     ]
    }
   ],
   "source": [
    "sentences  = [sent_tokenize(abstract) for abstract in matched_texts['Text'] ] # create tokenised list of cleaner abstracts\n",
    "matched_texts['Sentence'] = sentences                                   # copy that list back into df as a new column\n",
    "sentence_per_row = matched_texts.explode('Sentence')                    # explode column in new df with 1 row/sentence token\n",
    "print(\"How many sentences in total: \" + str(len(sentence_per_row)))     # check the length of new df\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "fd66d796",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "                                                                                                                                                                                                         Text  \\\n",
      "0     Despite the availability of most of the human genome sequence the accu rate identification of genes on the DNA sequence remains to be continu ously improved and updated. This procedure relies on t...   \n",
      "0     Despite the availability of most of the human genome sequence the accu rate identification of genes on the DNA sequence remains to be continu ously improved and updated. This procedure relies on t...   \n",
      "0     Despite the availability of most of the human genome sequence the accu rate identification of genes on the DNA sequence remains to be continu ously improved and updated. This procedure relies on t...   \n",
      "0     Despite the availability of most of the human genome sequence the accu rate identification of genes on the DNA sequence remains to be continu ously improved and updated. This procedure relies on t...   \n",
      "0     Despite the availability of most of the human genome sequence the accu rate identification of genes on the DNA sequence remains to be continu ously improved and updated. This procedure relies on t...   \n",
      "...                                                                                                                                                                                                       ...   \n",
      "3792  P0833Screening for ARX gene mutations is indicated for males problems. The study of this disorder has intensiÜed because with mental retardation associated with West syndrome and/or incidence is a...   \n",
      "3793  P0842Medium-chain acyl-CoA dehydrogenase deÜciency the localisation of further genes encoding amino acid transporter B. Z. Yang subunits within these regions is conceivable. One candidate is ASC- ...   \n",
      "3793  P0842Medium-chain acyl-CoA dehydrogenase deÜciency the localisation of further genes encoding amino acid transporter B. Z. Yang subunits within these regions is conceivable. One candidate is ASC- ...   \n",
      "3793  P0842Medium-chain acyl-CoA dehydrogenase deÜciency the localisation of further genes encoding amino acid transporter B. Z. Yang subunits within these regions is conceivable. One candidate is ASC- ...   \n",
      "3793  P0842Medium-chain acyl-CoA dehydrogenase deÜciency the localisation of further genes encoding amino acid transporter B. Z. Yang subunits within these regions is conceivable. One candidate is ASC- ...   \n",
      "\n",
      "                                                                                                                                                                       Sentence  \n",
      "0     Despite the availability of most of the human genome sequence the accu rate identification of genes on the DNA sequence remains to be continu ously improved and updated.  \n",
      "0                                                This procedure relies on the combined useof a variety of tools including similarity searches and exon/gene predictionprograms.  \n",
      "0                                           Similarity searches rely on a set of sequence data that has been increas ing very substantially and rapidly during the past months.  \n",
      "0                                                                                        These includehuman cDNAs mouse genomic DNA and cDNAs and puffer fish genomicsequences.  \n",
      "0                                                                      With these resources the definition of a vertebrate set ofgenes is progressively approaching completion.  \n",
      "...                                                                                                                                                                         ...  \n",
      "3792                                                                                                                                   The patients had elevated lactate levels  \n",
      "3793                                                    P0842Medium-chain acyl-CoA dehydrogenase deÜciency the localisation of further genes encoding amino acid transporter B.  \n",
      "3793                                                                                                                      Z. Yang subunits within these regions is conceivable.  \n",
      "3793                                                                                         One candidate is ASC- Baylor University Medical Center, Dallas, TX, United States.  \n",
      "3793                                                                                                                                                                      1 (SL  \n",
      "\n",
      "[56473 rows x 2 columns]\n"
     ]
    }
   ],
   "source": [
    "print(sentence_per_row[['Text','Sentence']])                            # have a look. The selected rows should have \n",
    "                                                                        # 'Text' the same, but 'Sentence' different "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "3795c76e",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "How many matching sentences: 7991\n"
     ]
    }
   ],
   "source": [
    "matched_sentences = sentence_per_row[sentence_per_row['Sentence'].str.contains('[Aa]utis|ASD|AS|[Aa]sperger')]\n",
    "                                                     # create a new data frame with only the sentences that contain keywords\n",
    "print(\"How many matching sentences: \" + str(len(matched_sentences)))            # check the length"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "ef8cdb17",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Now how many matching sentences: 7990\n"
     ]
    }
   ],
   "source": [
    "matched_sentences = matched_sentences[~matched_sentences['Sentence'].isnull()]  # remove any rows with empty 'Sentence' column\n",
    "matched_sentences = matched_sentences.drop_duplicates()                         # drop any duplicates\n",
    "print(\"Now how many matching sentences: \" + str(len(matched_sentences)))        # check length of remaining data frame"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "847a9ab3",
   "metadata": {},
   "source": [
    "In working with the matching sentences, it became clear there were several common errors, variations on how things were written and other annoying minor differences in the texts that made the manual checking more time-consuming than it needed to be.\n",
    "\n",
    "Further, the minor differences meant that the counting steps later on were counting \"child with ASD\" separately from \"child with autism\" when perhaps the more interesting distinction there is whether \"child with autism/ASD\" is more or less common than \"patient with autism/ASD\" or \"proband with autism/ASD\" or any other common person-nouns. \n",
    "\n",
    "Thus, this tidy_up_terminology function corrects several importing errors, spelling and style differences, and consolidates on terminology. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8e6a85b6",
   "metadata": {},
   "outputs": [],
   "source": [
    "def tidy_up_terminology (input):\n",
    "    space1 = re.sub(r'([A-Z]).(A-Z)', r'\\1. \\2', input)                 # removes multiple white spaces between words\n",
    "    space2 = re.sub(r'([a-z])(disorder|disability|spectrum)', r'\\1 \\2', space1) # adds a space in select run-ons\n",
    "    space3 = re.sub(r'([a-z])(disorder|spectrum)', r'\\1 \\2', space2)    # a second go at adding a space in select run-ons      \n",
    "    space4 = re.sub(r'(spec) (trum)', r'\\1\\2', space3)                  # removes a space between 'spec' and 'trum'\n",
    "    no_apost = re.sub(r'([Aa]sperger[\\S*?]s)', r'asperger', space4)     # lowercases, removes ' and S from '[Aa]sperger's' \n",
    "    lower1 = re.sub(r'Autis', r'autis', no_apost)                       # lowercases 'Autism' and 'Autistic'\n",
    "    lower2 = re.sub(r'[Aa]spergers|[Aa]sperger', r'asperger', lower1)   # lowercases/removes S from '[Aa]spergers' & '[Aa]sperger'\n",
    "    lower3 = re.sub(r'[Ss]pectrums|[Ss]pectra', r'spectrum', lower2)    # lowercases and removes various plurals for spectrum\n",
    "    lower4 = re.sub(r'[Ss]yndromes|[Ss]yndrome', r'syndrome', lower3)   # lowercases and removes plurals for syndrome\n",
    "    lower5 = re.sub(r'[Dd]isorders|Disorder', r'disorder', lower4)      # lowercases and removes plurals for disorder\n",
    "    lower6 = re.sub(r'[Dd]iseases|Disease', r'disease', lower5)         # lowercases and removes plurals for disease\n",
    "    plur = re.sub(r'ASDs', r'ASD', lower6)                              # removes plural from instances of more than one ASD\n",
    "    stan0 = re.sub(r'(autism|autistic|asperger) syndrome', r'autism spectrum', plur ) # turns select 'syndrome' to 'spectrum'\n",
    "    stan1 = re.sub(r'spectrum disease', r'spectrum disorder', stan0 )   # turns select 'disease' to 'disorder'\n",
    "    stan2 = re.sub(r'(autism|autistic|asperger) spectrum disorder \\(ASD\\)', r'ASD', stan1) # abbreviates various ASD definitions\n",
    "    stan3 = re.sub(r'(autism|autistic|asperger) spectrum disorder', r'ASD', stan2) # abbreviates various options to ASD\n",
    "    stan4 = re.sub(r'(autism|autistic|asperger) spectrum \\(AS\\)', r'ASD', stan3)  # standardises more options to ASD\n",
    "    stan5 = re.sub(r'(autism|autistic|asperger) spectrum', r'ASD', stan4)         # standardises more options to ASD\n",
    "    stan6 = re.sub(r'AS ', r'ASD ', stan5)                              # standardises 'AS ' to 'ASD ' - note trailing space\n",
    "    stan7 = re.sub(r'(autism|autistic|asperger) disorder', r'ASD', stan6) # abbreviates various ASD definitions\n",
    "    aut0 = re.sub(r'asperger autism', r'autism', stan7)                  # standardises 'asperger autism' to 'autism'\n",
    "    ID1 = re.sub(r'[Ii]ntellectual [Dd]isability \\(ID\\)', r'ID', aut0)\n",
    "    ID2 = re.sub(r'[Ii]ntellectual [Dd]isability', r'ID', ID1)\n",
    "\n",
    "    return(ID2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5fa88a1e",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5c95f250",
   "metadata": {},
   "outputs": [],
   "source": [
    " # Optional cell code block to test or understand what the tidy_up_terminology function does\n",
    "    \n",
    "tidy_test = \"Autism spectrum intellectual disability and autism ID, ASD \\\n",
    "            Autisticspectrum autisticspectrumdisorder ASD \\\n",
    "            Asperger's syndrome asperger's syndrome \\\n",
    "            intellectual disability Intellectual Disability (ID)\\\n",
    "            aspergers syndrome autism spectrum  ASDs ASD ID, and autism \"\n",
    "\n",
    "tidy_up_terminology(tidy_test)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d1708235",
   "metadata": {},
   "outputs": [],
   "source": [
    "tidy_text = [tidy_up_terminology(sentence) for sentence in matched_sentences['Sentence'] ] \n",
    "                                             # create abstract list without extra spaces/run-ons \n",
    "                                             # this is to improve sentence tokenisation later \n",
    "matched_sentences['Sentence'] = tidy_text    # copy the no extra space/run-on abstract list back into df as a new column"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d0fa94a6",
   "metadata": {},
   "outputs": [],
   "source": [
    "backup = matched_sentences                    # A backup is useful at this step because the next may not go the way you expect"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "08dd0c43",
   "metadata": {},
   "outputs": [],
   "source": [
    "matched_sentences = backup                    # If you need the backup, re-run this step. "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "49f0de07",
   "metadata": {},
   "source": [
    "## Extraction\n",
    "\n",
    "Following the cleaning phase, we move on to the extraction phase. This has two parts, first for the person-first extraction and then for the identity-first extraction. \n",
    "\n",
    "The results of both extractions are saved in their own column to make it easy to read and also to allow for a single sentence-token to contain both kinds of patterns. \n",
    "\n",
    "### Person-first pattern"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "64d739a6",
   "metadata": {},
   "outputs": [],
   "source": [
    "pattern_1 = [{\"POS\": \"NOUN\"},                                        # define the person-first pattern - start with a noun\n",
    "             {'DEP':'amod', 'OP':\"?\"},                               # followed by an optional modifier\n",
    "             {\"TEXT\": {\"REGEX\": \"(with|by|from)\"}},                  # followed by some words that set up the p-f pattern\n",
    "             {'DEP':'amod', 'OP':\"?\"},                               # then space for up to three optional modifiers\n",
    "             {'DEP':'amod', 'OP':\"?\"},\n",
    "             {'DEP':'amod', 'OP':\"?\"},\n",
    "             {\"TEXT\": {\"REGEX\": \"(^[Aa]utis|^[Aa]sperger|^ASD|^AS$)\"}}] # finally, the keywords (original format, just in case)\n",
    "\n",
    "# Matcher class object \n",
    "matcher = Matcher(nlp.vocab)                                         # define a matcher class object\n",
    "matcher.add(\"matching_1\", [pattern_1])                               # add my three person-first patterns to it\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "87d072ec",
   "metadata": {},
   "outputs": [],
   "source": [
    "def find_pattern_match(input):                                               # define a function that applies the person-first\n",
    "    thingy = nlp(input)                                                      # matcher class object to strings\n",
    "    match = matcher(thingy)                                                  # and returns any matches to the pattern(s)\n",
    "    if match == []:\n",
    "        out_value = ''\n",
    "    else:\n",
    "        hold_multi_spans = []\n",
    "        for match_id, start, end in match:\n",
    "                string_id = nlp.vocab.strings[match_id]  # Get string representation\n",
    "                span = thingy[start:end]  # The matched span\n",
    "                hold_multi_spans.append(span)\n",
    "        out_value = hold_multi_spans\n",
    "    return out_value"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "2ffd7373",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "7990"
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "matched_sentences['Person-first'] = matched_sentences.apply(lambda row: find_pattern_match(row.Sentence), axis = 1)\n",
    "                                                                        # apply the newly defined person-first matcher function\n",
    "                                                                        # and store the returned output in a new column\n",
    "len(matched_sentences)                                                  # double check length remains same"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "936cc982",
   "metadata": {},
   "source": [
    "### Identity-first pattern"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "dc18f3c5",
   "metadata": {},
   "outputs": [],
   "source": [
    "pattern_a = [{'DEP':'amod', 'OP':\"?\"},                                 # same for identity-first patterns,\n",
    "             {'DEP':'amod', 'OP':\"?\"},                                 # starting with two optional modifiers\n",
    "             {\"TEXT\": {\"REGEX\": \"(^[Aa]utis|^[Aa]sperger|^ASD|^AS$)\"}}, # the keywords (original format, just in case)\n",
    "             {'DEP':'amod', 'OP':\"?\"},                                 # then upt to three more optional modifiers\n",
    "             {'DEP':'amod', 'OP':\"?\"},\n",
    "             {'DEP':'amod', 'OP':\"?\"},\n",
    "             {\"POS\": \"NOUN\"}]                                          # and then a noun\n",
    "\n",
    "# Matcher class object                                         \n",
    "matcher = Matcher(nlp.vocab) \n",
    "matcher.add(\"matching_2\", [pattern_a])            # this overwrites the matcher object to identity-first"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "7f00defd",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "7990"
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "matched_sentences['Identity-first'] = matched_sentences.apply(lambda row: find_pattern_match(row.Sentence), axis = 1)\n",
    "                                                                        # apply the newly overwritten matcher function\n",
    "                                                                        # and store the returned output in a new column\n",
    "len(matched_sentences)                                                  # check the length - why not?"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "406db79a",
   "metadata": {},
   "source": [
    "### Consolidation\n",
    "\n",
    "Following the cleaning and extraction phases, the last phase is consolidation. This phase further refines the data by removing all the rows that do not contain a match for one or both of the patterns. For example, there would be a row for \"The child was tested for autism.\" because it contains a keyword of interest. However, this sentence would be eliminated in the consolidation phase as the keyword does not fit into either the person-first or identity-first patterns. \n",
    "\n",
    "Further, this phase goes on to lemmatise the extracted patterns so that they can be counted more easily. This phase also lowercases all occurrences of \"Autistic\", \"Autism\", and \"Asperger's\" as well as removing the apostrophe, the 's' and any non-white characters that might intrude between the 'r' and the 's' of \"Asperger's\". This phase also removes any square brackets, quotes and extra commas introduced by the lemmatisation process. \n",
    "\n",
    "This phase ends by writing out the consolidated data frame to a .csv for manual inspection. I could not find a feasible way of identifying whether or not the nouns matched in the extraction phase are person-nouns or not. As the list is not a totally unreasonable length (in the hundreds) I found it workable to \n",
    "* open in excel, \n",
    "* save the file under another name (e.g. pattern_matches_reviewed), \n",
    "* order the entire data set alphabetically by 'Person-first', \n",
    "* scan through the ordered results check whether each result in the 'Person-first' column is about a person, \n",
    "* removing entire rows if the 'Person-first' match is not about a person (checking the 'Sentence' or 'Text' column if needed)\n",
    "* re-order the entire data set alphabetically by 'Identity-first', \n",
    "* scan through the ordered results check whether each result in the 'Identity-first' column is about a person, \n",
    "* removing entire rows if the 'Identity-first' match is not about a person, \n",
    "* save file again. \n",
    "\n",
    "For example, 'association with autism' matches the person-first pattern but is not about a person, so this row was removed. Many more rows were removed in the 'Identity-first' matches as things like 'autistic behaviours' and 'autism testing' were removed for not being about people. \n",
    "\n",
    "NOTE: There were several instances of \"ASD dataset\" which are not easy to determine if they are about people or not. Do they mean dataset composed from blood tests taken as part of ASD testing? If so, each row in the data set would be a blood test with the possibility that more than one test comes from the same person. Or do they mean a pool of case records, each of which represents a single person? The former would not be \"about people\" but the second would. I did not remove these rows as we cannot be certain. Leaving them out would also have been a valid option, as long as the choice was clear. \n",
    "\n",
    "Coincidentally, during this manual checking part of the consolidation phase I learned that, in the context of human genetics research \"proband\" is a person-noun. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "b2d91446",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "954"
      ]
     },
     "execution_count": 18,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "matched_patterns = matched_sentences[(matched_sentences['Person-first'] != '') | (matched_sentences['Identity-first'] != '')]\n",
    "                                                     # keep only rows w/ non-null 'Person-first' and/or 'Identity-first' columns\n",
    "len(matched_patterns)                                # check length"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "58536fd7",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "957"
      ]
     },
     "execution_count": 19,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "matched_patterns = matched_patterns.explode('Person-first')    # explode 'Person-first' column to create 1 row per match\n",
    "                                                               # if there were two matches within the same sentence\n",
    "len(matched_patterns)                                          # check the length"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "80451426",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "1120"
      ]
     },
     "execution_count": 20,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "matched_patterns = matched_patterns.explode('Identity-first')  # Do the same for 'Identity-first' column\n",
    "len(matched_patterns)                                          # check the length"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "4ba5db63",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Unnamed: 0</th>\n",
       "      <th>Title</th>\n",
       "      <th>Session_Code</th>\n",
       "      <th>Authors_and_Affiliations</th>\n",
       "      <th>Text</th>\n",
       "      <th>Year</th>\n",
       "      <th>Email</th>\n",
       "      <th>Author</th>\n",
       "      <th>Affiliations</th>\n",
       "      <th>Sentence</th>\n",
       "      <th>Person-first</th>\n",
       "      <th>Identity-first</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>150</td>\n",
       "      <td>Location of the first predisposing gene locus for Asperger syndrome on chromosome 1q21 22</td>\n",
       "      <td>C102.</td>\n",
       "      <td>E. Jarvela1  T. Ylisaukko oja2  T. Nieminen3  E. Kempas1  M. Auranen1  L. Peltonen1 1National Public Health Institute  Helsinki  Finland  2National Public Health Insitute  Helsinki  Finland  3Uni...</td>\n",
       "      <td>autism was first described in 1944 by a Viennese physi cian Hans asperger who reported a group of boys with autism whose clinical features resembled autism with some modifications. Itis characteri...</td>\n",
       "      <td>2001.0</td>\n",
       "      <td>irma.jarvela@hus.fi</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>autism was first described in 1944 by a Viennese physi cian Hans asperger who reported a group of boys with autism whose clinical features resembled autism with some modifications.</td>\n",
       "      <td>(boys, with, autism)</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>150</td>\n",
       "      <td>Location of the first predisposing gene locus for Asperger syndrome on chromosome 1q21 22</td>\n",
       "      <td>C102.</td>\n",
       "      <td>E. Jarvela1  T. Ylisaukko oja2  T. Nieminen3  E. Kempas1  M. Auranen1  L. Peltonen1 1National Public Health Institute  Helsinki  Finland  2National Public Health Insitute  Helsinki  Finland  3Uni...</td>\n",
       "      <td>autism was first described in 1944 by a Viennese physi cian Hans asperger who reported a group of boys with autism whose clinical features resembled autism with some modifications. Itis characteri...</td>\n",
       "      <td>2001.0</td>\n",
       "      <td>irma.jarvela@hus.fi</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>We report the analysis of13 candidate gene loci associated with autism and schizophrenia in 17Finnish ASD families with autosomal dominant mode of inheritance.</td>\n",
       "      <td></td>\n",
       "      <td>(ASD, families)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>150</td>\n",
       "      <td>Location of the first predisposing gene locus for Asperger syndrome on chromosome 1q21 22</td>\n",
       "      <td>C102.</td>\n",
       "      <td>E. Jarvela1  T. Ylisaukko oja2  T. Nieminen3  E. Kempas1  M. Auranen1  L. Peltonen1 1National Public Health Institute  Helsinki  Finland  2National Public Health Insitute  Helsinki  Finland  3Uni...</td>\n",
       "      <td>autism was first described in 1944 by a Viennese physi cian Hans asperger who reported a group of boys with autism whose clinical features resembled autism with some modifications. Itis characteri...</td>\n",
       "      <td>2001.0</td>\n",
       "      <td>irma.jarvela@hus.fi</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Linkageto the previously reported predisposing loci for autism could not be repli cated with Finnish ASD families.</td>\n",
       "      <td></td>\n",
       "      <td>(Finnish, ASD, families)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>150</td>\n",
       "      <td>Location of the first predisposing gene locus for Asperger syndrome on chromosome 1q21 22</td>\n",
       "      <td>C102.</td>\n",
       "      <td>E. Jarvela1  T. Ylisaukko oja2  T. Nieminen3  E. Kempas1  M. Auranen1  L. Peltonen1 1National Public Health Institute  Helsinki  Finland  2National Public Health Insitute  Helsinki  Finland  3Uni...</td>\n",
       "      <td>autism was first described in 1944 by a Viennese physi cian Hans asperger who reported a group of boys with autism whose clinical features resembled autism with some modifications. Itis characteri...</td>\n",
       "      <td>2001.0</td>\n",
       "      <td>irma.jarvela@hus.fi</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Linkageto the previously reported predisposing loci for autism could not be repli cated with Finnish ASD families.</td>\n",
       "      <td></td>\n",
       "      <td>(ASD, families)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>400</td>\n",
       "      <td>Prader Willi and Angelman Syndromes in Chilean Patients. Clinical and Molecular Diagnosis.</td>\n",
       "      <td>P0295.</td>\n",
       "      <td>Curotto   L. Santa Mar a  A. Alliende  F. Cort s INTA University of Chile  SANTIAGO  C</td>\n",
       "      <td>Prader Willi (PWS) and Angelman AS) syndrome are multigenic disordercharacterized by developmental and neurobehavioral abnormalities. Dif ferent underlying genetic defects cause loss of expression...</td>\n",
       "      <td>2001.0</td>\n",
       "      <td>malliend@uec.inta.uchile.cl</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Approximately 70% of PWSand ASD patients have a deletion in 15q11 q13 whereas uniparental disomy(UPD) accounts for 25% of PWS cases and 2 3% of AS.</td>\n",
       "      <td></td>\n",
       "      <td>(ASD, patients)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3748</th>\n",
       "      <td>675</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>P0280Two cases of Cri Du Chat syndrome in the same family Spatial association of oppositely imprinted regions in late S-phase without a familial translocation or inversion. but not at other stages...</td>\n",
       "      <td>2004.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>and the PWS/ASD region on 15q11-13.</td>\n",
       "      <td></td>\n",
       "      <td>(ASD, region)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3748</th>\n",
       "      <td>675</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>P0280Two cases of Cri Du Chat syndrome in the same family Spatial association of oppositely imprinted regions in late S-phase without a familial translocation or inversion. but not at other stages...</td>\n",
       "      <td>2004.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>In lymphoblastoid cells of Here we present two cases with cri du chat syndrome who are a PWS patient with a complete deletion of the PWS/ASD region in second cousins.</td>\n",
       "      <td></td>\n",
       "      <td>(ASD, region)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3748</th>\n",
       "      <td>675</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>P0280Two cases of Cri Du Chat syndrome in the same family Spatial association of oppositely imprinted regions in late S-phase without a familial translocation or inversion. but not at other stages...</td>\n",
       "      <td>2004.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>The chromosome analysis with GTG the PWS/ASD locus but no ribosomal genes, the ARD between the two banding showed partial 5p deletion distal to 5p14.</td>\n",
       "      <td></td>\n",
       "      <td>(ASD, locus)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3776</th>\n",
       "      <td>1710</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>P0697How to prepare stable reference materials for genetic establish that treatment of cultured human dermal Übroblasts with testing several recombinant Übrillin-1 fragments induces up-regulation ...</td>\n",
       "      <td>2004.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>stable transfected human SK-N-ASD neuroblastoma cell lines A number of genetic CRMs are presently under development by a expressing ataxin-3 with a variable number of polyglutamine repeats: Europe...</td>\n",
       "      <td></td>\n",
       "      <td>(ASD, neuroblastoma)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3791</th>\n",
       "      <td>2157</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>P0835Presence of elevated lactate, lactate/pyruvate ratio and Analysis of the patientsÓclinical and molecular data demonstrated that acylcarnitine proÜle in patients with autism. all Üve patients ...</td>\n",
       "      <td>2004.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>P0835Presence of elevated lactate, lactate/pyruvate ratio and Analysis of the patientsÓclinical and molecular data demonstrated that acylcarnitine proÜle in patients with autism.</td>\n",
       "      <td>(patients, with, autism)</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1120 rows × 12 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      Unnamed: 0  \\\n",
       "8            150   \n",
       "8            150   \n",
       "8            150   \n",
       "8            150   \n",
       "21           400   \n",
       "...          ...   \n",
       "3748         675   \n",
       "3748         675   \n",
       "3748         675   \n",
       "3776        1710   \n",
       "3791        2157   \n",
       "\n",
       "                                                                                           Title  \\\n",
       "8     Location of the first predisposing gene locus for Asperger syndrome on chromosome 1q21 22    \n",
       "8     Location of the first predisposing gene locus for Asperger syndrome on chromosome 1q21 22    \n",
       "8     Location of the first predisposing gene locus for Asperger syndrome on chromosome 1q21 22    \n",
       "8     Location of the first predisposing gene locus for Asperger syndrome on chromosome 1q21 22    \n",
       "21    Prader Willi and Angelman Syndromes in Chilean Patients. Clinical and Molecular Diagnosis.   \n",
       "...                                                                                          ...   \n",
       "3748                                                                                         NaN   \n",
       "3748                                                                                         NaN   \n",
       "3748                                                                                         NaN   \n",
       "3776                                                                                         NaN   \n",
       "3791                                                                                         NaN   \n",
       "\n",
       "     Session_Code  \\\n",
       "8           C102.   \n",
       "8           C102.   \n",
       "8           C102.   \n",
       "8           C102.   \n",
       "21         P0295.   \n",
       "...           ...   \n",
       "3748          NaN   \n",
       "3748          NaN   \n",
       "3748          NaN   \n",
       "3776          NaN   \n",
       "3791          NaN   \n",
       "\n",
       "                                                                                                                                                                                     Authors_and_Affiliations  \\\n",
       "8      E. Jarvela1  T. Ylisaukko oja2  T. Nieminen3  E. Kempas1  M. Auranen1  L. Peltonen1 1National Public Health Institute  Helsinki  Finland  2National Public Health Insitute  Helsinki  Finland  3Uni...   \n",
       "8      E. Jarvela1  T. Ylisaukko oja2  T. Nieminen3  E. Kempas1  M. Auranen1  L. Peltonen1 1National Public Health Institute  Helsinki  Finland  2National Public Health Insitute  Helsinki  Finland  3Uni...   \n",
       "8      E. Jarvela1  T. Ylisaukko oja2  T. Nieminen3  E. Kempas1  M. Auranen1  L. Peltonen1 1National Public Health Institute  Helsinki  Finland  2National Public Health Insitute  Helsinki  Finland  3Uni...   \n",
       "8      E. Jarvela1  T. Ylisaukko oja2  T. Nieminen3  E. Kempas1  M. Auranen1  L. Peltonen1 1National Public Health Institute  Helsinki  Finland  2National Public Health Insitute  Helsinki  Finland  3Uni...   \n",
       "21                                                                                                                     Curotto   L. Santa Mar a  A. Alliende  F. Cort s INTA University of Chile  SANTIAGO  C   \n",
       "...                                                                                                                                                                                                       ...   \n",
       "3748                                                                                                                                                                                                      NaN   \n",
       "3748                                                                                                                                                                                                      NaN   \n",
       "3748                                                                                                                                                                                                      NaN   \n",
       "3776                                                                                                                                                                                                      NaN   \n",
       "3791                                                                                                                                                                                                      NaN   \n",
       "\n",
       "                                                                                                                                                                                                         Text  \\\n",
       "8     autism was first described in 1944 by a Viennese physi cian Hans asperger who reported a group of boys with autism whose clinical features resembled autism with some modifications. Itis characteri...   \n",
       "8     autism was first described in 1944 by a Viennese physi cian Hans asperger who reported a group of boys with autism whose clinical features resembled autism with some modifications. Itis characteri...   \n",
       "8     autism was first described in 1944 by a Viennese physi cian Hans asperger who reported a group of boys with autism whose clinical features resembled autism with some modifications. Itis characteri...   \n",
       "8     autism was first described in 1944 by a Viennese physi cian Hans asperger who reported a group of boys with autism whose clinical features resembled autism with some modifications. Itis characteri...   \n",
       "21    Prader Willi (PWS) and Angelman AS) syndrome are multigenic disordercharacterized by developmental and neurobehavioral abnormalities. Dif ferent underlying genetic defects cause loss of expression...   \n",
       "...                                                                                                                                                                                                       ...   \n",
       "3748  P0280Two cases of Cri Du Chat syndrome in the same family Spatial association of oppositely imprinted regions in late S-phase without a familial translocation or inversion. but not at other stages...   \n",
       "3748  P0280Two cases of Cri Du Chat syndrome in the same family Spatial association of oppositely imprinted regions in late S-phase without a familial translocation or inversion. but not at other stages...   \n",
       "3748  P0280Two cases of Cri Du Chat syndrome in the same family Spatial association of oppositely imprinted regions in late S-phase without a familial translocation or inversion. but not at other stages...   \n",
       "3776  P0697How to prepare stable reference materials for genetic establish that treatment of cultured human dermal Übroblasts with testing several recombinant Übrillin-1 fragments induces up-regulation ...   \n",
       "3791  P0835Presence of elevated lactate, lactate/pyruvate ratio and Analysis of the patientsÓclinical and molecular data demonstrated that acylcarnitine proÜle in patients with autism. all Üve patients ...   \n",
       "\n",
       "        Year                        Email Author Affiliations  \\\n",
       "8     2001.0          irma.jarvela@hus.fi    NaN          NaN   \n",
       "8     2001.0          irma.jarvela@hus.fi    NaN          NaN   \n",
       "8     2001.0          irma.jarvela@hus.fi    NaN          NaN   \n",
       "8     2001.0          irma.jarvela@hus.fi    NaN          NaN   \n",
       "21    2001.0  malliend@uec.inta.uchile.cl    NaN          NaN   \n",
       "...      ...                          ...    ...          ...   \n",
       "3748  2004.0                          NaN    NaN          NaN   \n",
       "3748  2004.0                          NaN    NaN          NaN   \n",
       "3748  2004.0                          NaN    NaN          NaN   \n",
       "3776  2004.0                          NaN    NaN          NaN   \n",
       "3791  2004.0                          NaN    NaN          NaN   \n",
       "\n",
       "                                                                                                                                                                                                     Sentence  \\\n",
       "8                        autism was first described in 1944 by a Viennese physi cian Hans asperger who reported a group of boys with autism whose clinical features resembled autism with some modifications.   \n",
       "8                                             We report the analysis of13 candidate gene loci associated with autism and schizophrenia in 17Finnish ASD families with autosomal dominant mode of inheritance.   \n",
       "8                                                                                          Linkageto the previously reported predisposing loci for autism could not be repli cated with Finnish ASD families.   \n",
       "8                                                                                          Linkageto the previously reported predisposing loci for autism could not be repli cated with Finnish ASD families.   \n",
       "21                                                        Approximately 70% of PWSand ASD patients have a deletion in 15q11 q13 whereas uniparental disomy(UPD) accounts for 25% of PWS cases and 2 3% of AS.   \n",
       "...                                                                                                                                                                                                       ...   \n",
       "3748                                                                                                                                                                      and the PWS/ASD region on 15q11-13.   \n",
       "3748                                   In lymphoblastoid cells of Here we present two cases with cri du chat syndrome who are a PWS patient with a complete deletion of the PWS/ASD region in second cousins.   \n",
       "3748                                                    The chromosome analysis with GTG the PWS/ASD locus but no ribosomal genes, the ARD between the two banding showed partial 5p deletion distal to 5p14.   \n",
       "3776  stable transfected human SK-N-ASD neuroblastoma cell lines A number of genetic CRMs are presently under development by a expressing ataxin-3 with a variable number of polyglutamine repeats: Europe...   \n",
       "3791                       P0835Presence of elevated lactate, lactate/pyruvate ratio and Analysis of the patientsÓclinical and molecular data demonstrated that acylcarnitine proÜle in patients with autism.   \n",
       "\n",
       "                  Person-first            Identity-first  \n",
       "8         (boys, with, autism)                            \n",
       "8                                        (ASD, families)  \n",
       "8                               (Finnish, ASD, families)  \n",
       "8                                        (ASD, families)  \n",
       "21                                       (ASD, patients)  \n",
       "...                        ...                       ...  \n",
       "3748                                       (ASD, region)  \n",
       "3748                                       (ASD, region)  \n",
       "3748                                        (ASD, locus)  \n",
       "3776                                (ASD, neuroblastoma)  \n",
       "3791  (patients, with, autism)                            \n",
       "\n",
       "[1120 rows x 12 columns]"
      ]
     },
     "execution_count": 21,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "matched_patterns                                               # have a look at them"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "id": "d9e44521",
   "metadata": {},
   "outputs": [],
   "source": [
    "Lem = WordNetLemmatizer()                         # Define a short way to call the WordNetLemmatizer\n",
    "\n",
    "def consolidate_matched_patterns (input):         # \n",
    "    final_lemma_list = []\n",
    "    temp_lemma_list = []\n",
    "    for phrase in input:                       # start for loop looking at each pattern in the person-first pattern column\n",
    "        phrase_as_string = str(phrase)                               # hold the current pattern\n",
    "        words_in_phrase = phrase_as_string.split() # split the current pattern into words\n",
    "        for word in words_in_phrase :                            # for each word in the split up words\n",
    "            lemma = Lem.lemmatize(word)             # turn that word into a lemma\n",
    "            temp_lemma_list.append(lemma)                # append that lemma to a temporary list\n",
    "        string_lem = str(temp_lemma_list)              # turn that temporary list into a string\n",
    "        stripped_lem = re.sub(r\"\\[|\\]|\\'|\\,\",'', string_lem)  # remove  square brackets, commas and '' marks from the string\n",
    "        final_lemma_list.append(stripped_lem)        # append the string version of the list to the output list\n",
    "        temp_lemma_list = []                               # ensure the temp variable is empty\n",
    "\n",
    "    return(final_lemma_list)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "id": "90b670fe",
   "metadata": {},
   "outputs": [],
   "source": [
    "person_lemma_list = consolidate_matched_patterns(matched_patterns['Person-first'])\n",
    "identity_lemma_list = consolidate_matched_patterns(matched_patterns['Identity-first'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "id": "ad90adb0",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Unnamed: 0</th>\n",
       "      <th>Title</th>\n",
       "      <th>Session_Code</th>\n",
       "      <th>Authors_and_Affiliations</th>\n",
       "      <th>Text</th>\n",
       "      <th>Year</th>\n",
       "      <th>Email</th>\n",
       "      <th>Author</th>\n",
       "      <th>Affiliations</th>\n",
       "      <th>Sentence</th>\n",
       "      <th>Person-first</th>\n",
       "      <th>Identity-first</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>150</td>\n",
       "      <td>Location of the first predisposing gene locus for Asperger syndrome on chromosome 1q21 22</td>\n",
       "      <td>C102.</td>\n",
       "      <td>E. Jarvela1  T. Ylisaukko oja2  T. Nieminen3  E. Kempas1  M. Auranen1  L. Peltonen1 1National Public Health Institute  Helsinki  Finland  2National Public Health Insitute  Helsinki  Finland  3Uni...</td>\n",
       "      <td>autism was first described in 1944 by a Viennese physi cian Hans asperger who reported a group of boys with autism whose clinical features resembled autism with some modifications. Itis characteri...</td>\n",
       "      <td>2001.0</td>\n",
       "      <td>irma.jarvela@hus.fi</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>autism was first described in 1944 by a Viennese physi cian Hans asperger who reported a group of boys with autism whose clinical features resembled autism with some modifications.</td>\n",
       "      <td>boy with autism</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>150</td>\n",
       "      <td>Location of the first predisposing gene locus for Asperger syndrome on chromosome 1q21 22</td>\n",
       "      <td>C102.</td>\n",
       "      <td>E. Jarvela1  T. Ylisaukko oja2  T. Nieminen3  E. Kempas1  M. Auranen1  L. Peltonen1 1National Public Health Institute  Helsinki  Finland  2National Public Health Insitute  Helsinki  Finland  3Uni...</td>\n",
       "      <td>autism was first described in 1944 by a Viennese physi cian Hans asperger who reported a group of boys with autism whose clinical features resembled autism with some modifications. Itis characteri...</td>\n",
       "      <td>2001.0</td>\n",
       "      <td>irma.jarvela@hus.fi</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>We report the analysis of13 candidate gene loci associated with autism and schizophrenia in 17Finnish ASD families with autosomal dominant mode of inheritance.</td>\n",
       "      <td></td>\n",
       "      <td>ASD family</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>150</td>\n",
       "      <td>Location of the first predisposing gene locus for Asperger syndrome on chromosome 1q21 22</td>\n",
       "      <td>C102.</td>\n",
       "      <td>E. Jarvela1  T. Ylisaukko oja2  T. Nieminen3  E. Kempas1  M. Auranen1  L. Peltonen1 1National Public Health Institute  Helsinki  Finland  2National Public Health Insitute  Helsinki  Finland  3Uni...</td>\n",
       "      <td>autism was first described in 1944 by a Viennese physi cian Hans asperger who reported a group of boys with autism whose clinical features resembled autism with some modifications. Itis characteri...</td>\n",
       "      <td>2001.0</td>\n",
       "      <td>irma.jarvela@hus.fi</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Linkageto the previously reported predisposing loci for autism could not be repli cated with Finnish ASD families.</td>\n",
       "      <td></td>\n",
       "      <td>Finnish ASD family</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>150</td>\n",
       "      <td>Location of the first predisposing gene locus for Asperger syndrome on chromosome 1q21 22</td>\n",
       "      <td>C102.</td>\n",
       "      <td>E. Jarvela1  T. Ylisaukko oja2  T. Nieminen3  E. Kempas1  M. Auranen1  L. Peltonen1 1National Public Health Institute  Helsinki  Finland  2National Public Health Insitute  Helsinki  Finland  3Uni...</td>\n",
       "      <td>autism was first described in 1944 by a Viennese physi cian Hans asperger who reported a group of boys with autism whose clinical features resembled autism with some modifications. Itis characteri...</td>\n",
       "      <td>2001.0</td>\n",
       "      <td>irma.jarvela@hus.fi</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Linkageto the previously reported predisposing loci for autism could not be repli cated with Finnish ASD families.</td>\n",
       "      <td></td>\n",
       "      <td>ASD family</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>400</td>\n",
       "      <td>Prader Willi and Angelman Syndromes in Chilean Patients. Clinical and Molecular Diagnosis.</td>\n",
       "      <td>P0295.</td>\n",
       "      <td>Curotto   L. Santa Mar a  A. Alliende  F. Cort s INTA University of Chile  SANTIAGO  C</td>\n",
       "      <td>Prader Willi (PWS) and Angelman AS) syndrome are multigenic disordercharacterized by developmental and neurobehavioral abnormalities. Dif ferent underlying genetic defects cause loss of expression...</td>\n",
       "      <td>2001.0</td>\n",
       "      <td>malliend@uec.inta.uchile.cl</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Approximately 70% of PWSand ASD patients have a deletion in 15q11 q13 whereas uniparental disomy(UPD) accounts for 25% of PWS cases and 2 3% of AS.</td>\n",
       "      <td></td>\n",
       "      <td>ASD patient</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3748</th>\n",
       "      <td>675</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>P0280Two cases of Cri Du Chat syndrome in the same family Spatial association of oppositely imprinted regions in late S-phase without a familial translocation or inversion. but not at other stages...</td>\n",
       "      <td>2004.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>and the PWS/ASD region on 15q11-13.</td>\n",
       "      <td></td>\n",
       "      <td>ASD region</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3748</th>\n",
       "      <td>675</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>P0280Two cases of Cri Du Chat syndrome in the same family Spatial association of oppositely imprinted regions in late S-phase without a familial translocation or inversion. but not at other stages...</td>\n",
       "      <td>2004.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>In lymphoblastoid cells of Here we present two cases with cri du chat syndrome who are a PWS patient with a complete deletion of the PWS/ASD region in second cousins.</td>\n",
       "      <td></td>\n",
       "      <td>ASD region</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3748</th>\n",
       "      <td>675</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>P0280Two cases of Cri Du Chat syndrome in the same family Spatial association of oppositely imprinted regions in late S-phase without a familial translocation or inversion. but not at other stages...</td>\n",
       "      <td>2004.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>The chromosome analysis with GTG the PWS/ASD locus but no ribosomal genes, the ARD between the two banding showed partial 5p deletion distal to 5p14.</td>\n",
       "      <td></td>\n",
       "      <td>ASD locus</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3776</th>\n",
       "      <td>1710</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>P0697How to prepare stable reference materials for genetic establish that treatment of cultured human dermal Übroblasts with testing several recombinant Übrillin-1 fragments induces up-regulation ...</td>\n",
       "      <td>2004.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>stable transfected human SK-N-ASD neuroblastoma cell lines A number of genetic CRMs are presently under development by a expressing ataxin-3 with a variable number of polyglutamine repeats: Europe...</td>\n",
       "      <td></td>\n",
       "      <td>ASD neuroblastoma</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3791</th>\n",
       "      <td>2157</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>P0835Presence of elevated lactate, lactate/pyruvate ratio and Analysis of the patientsÓclinical and molecular data demonstrated that acylcarnitine proÜle in patients with autism. all Üve patients ...</td>\n",
       "      <td>2004.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>P0835Presence of elevated lactate, lactate/pyruvate ratio and Analysis of the patientsÓclinical and molecular data demonstrated that acylcarnitine proÜle in patients with autism.</td>\n",
       "      <td>patient with autism</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1109 rows × 12 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      Unnamed: 0  \\\n",
       "8            150   \n",
       "8            150   \n",
       "8            150   \n",
       "8            150   \n",
       "21           400   \n",
       "...          ...   \n",
       "3748         675   \n",
       "3748         675   \n",
       "3748         675   \n",
       "3776        1710   \n",
       "3791        2157   \n",
       "\n",
       "                                                                                           Title  \\\n",
       "8     Location of the first predisposing gene locus for Asperger syndrome on chromosome 1q21 22    \n",
       "8     Location of the first predisposing gene locus for Asperger syndrome on chromosome 1q21 22    \n",
       "8     Location of the first predisposing gene locus for Asperger syndrome on chromosome 1q21 22    \n",
       "8     Location of the first predisposing gene locus for Asperger syndrome on chromosome 1q21 22    \n",
       "21    Prader Willi and Angelman Syndromes in Chilean Patients. Clinical and Molecular Diagnosis.   \n",
       "...                                                                                          ...   \n",
       "3748                                                                                         NaN   \n",
       "3748                                                                                         NaN   \n",
       "3748                                                                                         NaN   \n",
       "3776                                                                                         NaN   \n",
       "3791                                                                                         NaN   \n",
       "\n",
       "     Session_Code  \\\n",
       "8           C102.   \n",
       "8           C102.   \n",
       "8           C102.   \n",
       "8           C102.   \n",
       "21         P0295.   \n",
       "...           ...   \n",
       "3748          NaN   \n",
       "3748          NaN   \n",
       "3748          NaN   \n",
       "3776          NaN   \n",
       "3791          NaN   \n",
       "\n",
       "                                                                                                                                                                                     Authors_and_Affiliations  \\\n",
       "8      E. Jarvela1  T. Ylisaukko oja2  T. Nieminen3  E. Kempas1  M. Auranen1  L. Peltonen1 1National Public Health Institute  Helsinki  Finland  2National Public Health Insitute  Helsinki  Finland  3Uni...   \n",
       "8      E. Jarvela1  T. Ylisaukko oja2  T. Nieminen3  E. Kempas1  M. Auranen1  L. Peltonen1 1National Public Health Institute  Helsinki  Finland  2National Public Health Insitute  Helsinki  Finland  3Uni...   \n",
       "8      E. Jarvela1  T. Ylisaukko oja2  T. Nieminen3  E. Kempas1  M. Auranen1  L. Peltonen1 1National Public Health Institute  Helsinki  Finland  2National Public Health Insitute  Helsinki  Finland  3Uni...   \n",
       "8      E. Jarvela1  T. Ylisaukko oja2  T. Nieminen3  E. Kempas1  M. Auranen1  L. Peltonen1 1National Public Health Institute  Helsinki  Finland  2National Public Health Insitute  Helsinki  Finland  3Uni...   \n",
       "21                                                                                                                     Curotto   L. Santa Mar a  A. Alliende  F. Cort s INTA University of Chile  SANTIAGO  C   \n",
       "...                                                                                                                                                                                                       ...   \n",
       "3748                                                                                                                                                                                                      NaN   \n",
       "3748                                                                                                                                                                                                      NaN   \n",
       "3748                                                                                                                                                                                                      NaN   \n",
       "3776                                                                                                                                                                                                      NaN   \n",
       "3791                                                                                                                                                                                                      NaN   \n",
       "\n",
       "                                                                                                                                                                                                         Text  \\\n",
       "8     autism was first described in 1944 by a Viennese physi cian Hans asperger who reported a group of boys with autism whose clinical features resembled autism with some modifications. Itis characteri...   \n",
       "8     autism was first described in 1944 by a Viennese physi cian Hans asperger who reported a group of boys with autism whose clinical features resembled autism with some modifications. Itis characteri...   \n",
       "8     autism was first described in 1944 by a Viennese physi cian Hans asperger who reported a group of boys with autism whose clinical features resembled autism with some modifications. Itis characteri...   \n",
       "8     autism was first described in 1944 by a Viennese physi cian Hans asperger who reported a group of boys with autism whose clinical features resembled autism with some modifications. Itis characteri...   \n",
       "21    Prader Willi (PWS) and Angelman AS) syndrome are multigenic disordercharacterized by developmental and neurobehavioral abnormalities. Dif ferent underlying genetic defects cause loss of expression...   \n",
       "...                                                                                                                                                                                                       ...   \n",
       "3748  P0280Two cases of Cri Du Chat syndrome in the same family Spatial association of oppositely imprinted regions in late S-phase without a familial translocation or inversion. but not at other stages...   \n",
       "3748  P0280Two cases of Cri Du Chat syndrome in the same family Spatial association of oppositely imprinted regions in late S-phase without a familial translocation or inversion. but not at other stages...   \n",
       "3748  P0280Two cases of Cri Du Chat syndrome in the same family Spatial association of oppositely imprinted regions in late S-phase without a familial translocation or inversion. but not at other stages...   \n",
       "3776  P0697How to prepare stable reference materials for genetic establish that treatment of cultured human dermal Übroblasts with testing several recombinant Übrillin-1 fragments induces up-regulation ...   \n",
       "3791  P0835Presence of elevated lactate, lactate/pyruvate ratio and Analysis of the patientsÓclinical and molecular data demonstrated that acylcarnitine proÜle in patients with autism. all Üve patients ...   \n",
       "\n",
       "        Year                        Email Author Affiliations  \\\n",
       "8     2001.0          irma.jarvela@hus.fi    NaN          NaN   \n",
       "8     2001.0          irma.jarvela@hus.fi    NaN          NaN   \n",
       "8     2001.0          irma.jarvela@hus.fi    NaN          NaN   \n",
       "8     2001.0          irma.jarvela@hus.fi    NaN          NaN   \n",
       "21    2001.0  malliend@uec.inta.uchile.cl    NaN          NaN   \n",
       "...      ...                          ...    ...          ...   \n",
       "3748  2004.0                          NaN    NaN          NaN   \n",
       "3748  2004.0                          NaN    NaN          NaN   \n",
       "3748  2004.0                          NaN    NaN          NaN   \n",
       "3776  2004.0                          NaN    NaN          NaN   \n",
       "3791  2004.0                          NaN    NaN          NaN   \n",
       "\n",
       "                                                                                                                                                                                                     Sentence  \\\n",
       "8                        autism was first described in 1944 by a Viennese physi cian Hans asperger who reported a group of boys with autism whose clinical features resembled autism with some modifications.   \n",
       "8                                             We report the analysis of13 candidate gene loci associated with autism and schizophrenia in 17Finnish ASD families with autosomal dominant mode of inheritance.   \n",
       "8                                                                                          Linkageto the previously reported predisposing loci for autism could not be repli cated with Finnish ASD families.   \n",
       "8                                                                                          Linkageto the previously reported predisposing loci for autism could not be repli cated with Finnish ASD families.   \n",
       "21                                                        Approximately 70% of PWSand ASD patients have a deletion in 15q11 q13 whereas uniparental disomy(UPD) accounts for 25% of PWS cases and 2 3% of AS.   \n",
       "...                                                                                                                                                                                                       ...   \n",
       "3748                                                                                                                                                                      and the PWS/ASD region on 15q11-13.   \n",
       "3748                                   In lymphoblastoid cells of Here we present two cases with cri du chat syndrome who are a PWS patient with a complete deletion of the PWS/ASD region in second cousins.   \n",
       "3748                                                    The chromosome analysis with GTG the PWS/ASD locus but no ribosomal genes, the ARD between the two banding showed partial 5p deletion distal to 5p14.   \n",
       "3776  stable transfected human SK-N-ASD neuroblastoma cell lines A number of genetic CRMs are presently under development by a expressing ataxin-3 with a variable number of polyglutamine repeats: Europe...   \n",
       "3791                       P0835Presence of elevated lactate, lactate/pyruvate ratio and Analysis of the patientsÓclinical and molecular data demonstrated that acylcarnitine proÜle in patients with autism.   \n",
       "\n",
       "             Person-first      Identity-first  \n",
       "8         boy with autism                      \n",
       "8                                  ASD family  \n",
       "8                          Finnish ASD family  \n",
       "8                                  ASD family  \n",
       "21                                ASD patient  \n",
       "...                   ...                 ...  \n",
       "3748                               ASD region  \n",
       "3748                               ASD region  \n",
       "3748                                ASD locus  \n",
       "3776                        ASD neuroblastoma  \n",
       "3791  patient with autism                      \n",
       "\n",
       "[1109 rows x 12 columns]"
      ]
     },
     "execution_count": 24,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "matched_patterns['Person-first'] = person_lemma_list    # copy the person-first output to new column in data frame \n",
    "matched_patterns['Identity-first'] = identity_lemma_list  # copy the identity-first output to new column in data frame \n",
    "matched_patterns = matched_patterns.drop_duplicates()                         # drop any duplicates\n",
    "matched_patterns                                                   # have a look at the data frame with its new columns"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "id": "20ac2138",
   "metadata": {},
   "outputs": [],
   "source": [
    "matched_patterns.to_csv('..\\\\output\\\\pattern_matches_to_review.csv')        \n",
    "                                                            # Write the data frame to a .csv for manual processing in excel"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9897ba53",
   "metadata": {},
   "source": [
    "At this point, I open the file in Excel (for example), removed the brackets, quotation marks and commas in the Person-first lemmatised and Identity-first lemmatised columns, then sort by each of one of these columns. I then scan through the results, removing any rows that are obviously not about people (e.g. \"autistic testing\") and checking the 'Text' column on any that are unclear 'autistic quartets'). I then sort by the other column and repeat the step of reviewing and deleting non-person rows. Save under \"pattern_matches_reviewed.csv\" for the next step. "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "35d2281a",
   "metadata": {},
   "source": [
    "## Exlporing statistics for person-first and identity-first"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "388d8be8",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Unnamed: 0.1</th>\n",
       "      <th>Unnamed: 0</th>\n",
       "      <th>Title</th>\n",
       "      <th>Session_Code</th>\n",
       "      <th>Authors_and_Affiliations</th>\n",
       "      <th>Text</th>\n",
       "      <th>Year</th>\n",
       "      <th>Email</th>\n",
       "      <th>Author</th>\n",
       "      <th>Affiliations</th>\n",
       "      <th>Sentence</th>\n",
       "      <th>Person-first</th>\n",
       "      <th>Identity-first</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>2012</td>\n",
       "      <td>134</td>\n",
       "      <td>Whole genome sequencing reveals the mutation characteristics in Autism Spectrum Disorder</td>\n",
       "      <td>C10.1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>JohnÃ¢ÂÂs, NL, Canada. autism is genetically heterogeneous, with evi dence for hundreds of susceptibility loci. Previous microarray and exome sequencing studies have examined portions of the gen...</td>\n",
       "      <td>2015</td>\n",
       "      <td>NaN</td>\n",
       "      <td>R. K. Yuen1, B. Thiruvahindrapuram1, D. Merico1, S. Walker1, K. Tammimies1,2, N. Hoang1,  C. Chrysler3, T . Nalpathamkalam1, G. Pellecchia1, Y. Liu1,4, M. J. Gazzellone1, L. DÃ¢ÂÂ Abate1, E.  De...</td>\n",
       "      <td>1The Hospital for Sick Children, Toronto, ON, Canada, 2Karolinska Institutet, Stockholm,  Sweden, 3McMaster University, Hamilton, ON, Canada, 4Qilu ChildrenÃ¢ÂÂs Hospital of  Shandong University...</td>\n",
       "      <td>We have now sequenced 200 additional ASD families using WGS.</td>\n",
       "      <td>NaN</td>\n",
       "      <td>additional ASD family</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1504</td>\n",
       "      <td>728</td>\n",
       "      <td>The association analysis of ABHD11 gene (rs2293484 and rs10279013 with autism in a South African population.</td>\n",
       "      <td>P06.05</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Background : autism is genetically inherited neurodevelopmental disorder characterized by significant impairments of social interaction, difficulties in communication and displays of restrictive o...</td>\n",
       "      <td>2013</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Z. Arieff1, J. R. Sharma1, M. Davids1, M. Kaur2</td>\n",
       "      <td>1Department of Biotechnogy,University of Western Cape, Private Bag X17, Bellville 7535,  South Africa, South Africa, 2Computational Bioscience Research Center, King Abdullah  University of Science...</td>\n",
       "      <td>Aims : The aim of this study was to investigate the association of two SNPs from genes ABHD11 (rs2293484 and rs10279013) in South African autistic population.</td>\n",
       "      <td>NaN</td>\n",
       "      <td>African autistic population</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1811</td>\n",
       "      <td>905</td>\n",
       "      <td>MThe Haplotype analysis to test for the association of three SNPs with autism in a South African Population</td>\n",
       "      <td>P09.152</td>\n",
       "      <td>NaN</td>\n",
       "      <td>autism is a neuropsychiatric developmental disorder characterized by communication difficulties and impaired social interac tion. The anhydrolase domain containing 11 gene (ABHD11 ) is located on ...</td>\n",
       "      <td>2014</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Z. Arieff1, J. R. Sharma1, M. Davids1, M. Kaur2</td>\n",
       "      <td>1Department of Biotechnology, University of Western Cape, Bellville, South Africa,  2Computational Bioscience Research Center, King Abdullah University of Science and  Technology, Thuwal, Saudi Ar...</td>\n",
       "      <td>The aim of this study was to investigate the association of SNPs from genes ABHD11 (rs2293484 and rs10279013) and ABHD13 (rs17060) with haplotype ana lysis in South African autistic population.</td>\n",
       "      <td>NaN</td>\n",
       "      <td>African autistic population</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   Unnamed: 0.1  Unnamed: 0  \\\n",
       "0          2012         134   \n",
       "1          1504         728   \n",
       "2          1811         905   \n",
       "\n",
       "                                                                                                          Title  \\\n",
       "0                      Whole genome sequencing reveals the mutation characteristics in Autism Spectrum Disorder   \n",
       "1  The association analysis of ABHD11 gene (rs2293484 and rs10279013 with autism in a South African population.   \n",
       "2   MThe Haplotype analysis to test for the association of three SNPs with autism in a South African Population   \n",
       "\n",
       "  Session_Code Authors_and_Affiliations  \\\n",
       "0        C10.1                      NaN   \n",
       "1       P06.05                      NaN   \n",
       "2      P09.152                      NaN   \n",
       "\n",
       "                                                                                                                                                                                                      Text  \\\n",
       "0  JohnÃ¢ÂÂs, NL, Canada. autism is genetically heterogeneous, with evi dence for hundreds of susceptibility loci. Previous microarray and exome sequencing studies have examined portions of the gen...   \n",
       "1  Background : autism is genetically inherited neurodevelopmental disorder characterized by significant impairments of social interaction, difficulties in communication and displays of restrictive o...   \n",
       "2  autism is a neuropsychiatric developmental disorder characterized by communication difficulties and impaired social interac tion. The anhydrolase domain containing 11 gene (ABHD11 ) is located on ...   \n",
       "\n",
       "   Year Email  \\\n",
       "0  2015   NaN   \n",
       "1  2013   NaN   \n",
       "2  2014   NaN   \n",
       "\n",
       "                                                                                                                                                                                                    Author  \\\n",
       "0  R. K. Yuen1, B. Thiruvahindrapuram1, D. Merico1, S. Walker1, K. Tammimies1,2, N. Hoang1,  C. Chrysler3, T . Nalpathamkalam1, G. Pellecchia1, Y. Liu1,4, M. J. Gazzellone1, L. DÃ¢ÂÂ Abate1, E.  De...   \n",
       "1                                                                                                                                                          Z. Arieff1, J. R. Sharma1, M. Davids1, M. Kaur2   \n",
       "2                                                                                                                                                          Z. Arieff1, J. R. Sharma1, M. Davids1, M. Kaur2   \n",
       "\n",
       "                                                                                                                                                                                              Affiliations  \\\n",
       "0  1The Hospital for Sick Children, Toronto, ON, Canada, 2Karolinska Institutet, Stockholm,  Sweden, 3McMaster University, Hamilton, ON, Canada, 4Qilu ChildrenÃ¢ÂÂs Hospital of  Shandong University...   \n",
       "1  1Department of Biotechnogy,University of Western Cape, Private Bag X17, Bellville 7535,  South Africa, South Africa, 2Computational Bioscience Research Center, King Abdullah  University of Science...   \n",
       "2  1Department of Biotechnology, University of Western Cape, Bellville, South Africa,  2Computational Bioscience Research Center, King Abdullah University of Science and  Technology, Thuwal, Saudi Ar...   \n",
       "\n",
       "                                                                                                                                                                                            Sentence  \\\n",
       "0                                                                                                                                       We have now sequenced 200 additional ASD families using WGS.   \n",
       "1                                     Aims : The aim of this study was to investigate the association of two SNPs from genes ABHD11 (rs2293484 and rs10279013) in South African autistic population.   \n",
       "2  The aim of this study was to investigate the association of SNPs from genes ABHD11 (rs2293484 and rs10279013) and ABHD13 (rs17060) with haplotype ana lysis in South African autistic population.   \n",
       "\n",
       "  Person-first               Identity-first  \n",
       "0          NaN        additional ASD family  \n",
       "1          NaN  African autistic population  \n",
       "2          NaN  African autistic population  "
      ]
     },
     "execution_count": 19,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "reviewed_matches = pd.read_csv('..\\\\output\\\\pattern_matches_reviewed.csv')    # one for just those that match the keyword\n",
    "reviewed_matches.head(3)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "81fbac29",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "There are 545 rows in the post-manual review data frame coming from 250 unique titles.\n"
     ]
    }
   ],
   "source": [
    "print(\"There are \" + \n",
    "      str(len(reviewed_matches)) + \" rows in the post-manual review data frame coming from \" +\n",
    "      str(reviewed_matches['Title'].nunique()) +\n",
    "      \" unique titles.\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "aa2c9802",
   "metadata": {},
   "outputs": [],
   "source": [
    "def find_PF_nouns(input):\n",
    "    output = []\n",
    "    for thingy in input:\n",
    "        if isinstance(thingy,str):\n",
    "            word_list = thingy.split()\n",
    "            noun = word_list[0]\n",
    "            output.append(noun)\n",
    "        else:\n",
    "            output.append(\"\")\n",
    "    return output\n",
    "\n",
    "def find_IF_nouns(input):\n",
    "    output = []\n",
    "    for thingy in input:\n",
    "        if isinstance(thingy,str):\n",
    "            word_list = thingy.split()\n",
    "            noun = word_list[-1]\n",
    "            output.append(noun)\n",
    "        else:\n",
    "            output.append(\"\")\n",
    "    return output"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "id": "6eda37d9",
   "metadata": {},
   "outputs": [],
   "source": [
    "reviewed_matches['PF_nouns'] = find_PF_nouns(reviewed_matches['Person-first'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "id": "8b942411",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "In total, there are 219 examples of PFL, coming from 138 unique titles, with 55 unique patterns and 23 unique nouns.\n"
     ]
    }
   ],
   "source": [
    "print(\"In total, there are \" + str(reviewed_matches['Person-first'].count()) +\n",
    "      \" examples of PFL, coming from \" +\n",
    "      str(len(reviewed_matches[reviewed_matches[\"Person-first\"].notnull() == True].groupby(['Title']).nunique())) +\n",
    "      \" unique titles, with \" +\n",
    "      str(reviewed_matches['Person-first'].nunique()) + \" unique patterns and \" +\n",
    "      str(reviewed_matches['PF_nouns'].nunique()) + \" unique nouns.\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "id": "1028fa3d",
   "metadata": {},
   "outputs": [],
   "source": [
    "reviewed_matches['IF_nouns'] = find_IF_nouns(reviewed_matches['Identity-first'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "id": "9e7e954e",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "In total, there are 331 examples of IFL, coming from 158 unique titles, with 94 unique patterns and 24 unique nouns.\n"
     ]
    }
   ],
   "source": [
    "print(\"In total, there are \" + str(reviewed_matches['Identity-first'].count()) +\n",
    "      \" examples of IFL, coming from \" +\n",
    "      str(len(reviewed_matches[reviewed_matches[\"Identity-first\"].notnull() == True].groupby(['Title']).nunique())) +\n",
    "      \" unique titles, with \" +\n",
    "      str(reviewed_matches['Identity-first'].nunique()) + \" unique patterns and \" +\n",
    "      str(reviewed_matches['IF_nouns'].nunique()) + \" unique nouns.\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 45,
   "id": "ebf88c1a",
   "metadata": {},
   "outputs": [
    {
     "ename": "AttributeError",
     "evalue": "'int' object has no attribute 'sort_values'",
     "output_type": "error",
     "traceback": [
      "\u001b[1;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[1;31mAttributeError\u001b[0m                            Traceback (most recent call last)",
      "Cell \u001b[1;32mIn[45], line 1\u001b[0m\n\u001b[1;32m----> 1\u001b[0m \u001b[38;5;28mprint\u001b[39m(\u001b[43mreviewed_matches\u001b[49m\u001b[43m[\u001b[49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[38;5;124;43mPerson-first\u001b[39;49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[43m]\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mnunique\u001b[49m\u001b[43m(\u001b[49m\u001b[43m)\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43msort_values\u001b[49m(ascending\u001b[38;5;241m=\u001b[39m\u001b[38;5;28;01mFalse\u001b[39;00m)\u001b[38;5;241m.\u001b[39mhead(\u001b[38;5;241m10\u001b[39m))\n\u001b[0;32m      2\u001b[0m \u001b[38;5;28mprint\u001b[39m(\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124m\"\u001b[39m)\n\u001b[0;32m      3\u001b[0m \u001b[38;5;28mprint\u001b[39m(reviewed_matches\u001b[38;5;241m.\u001b[39mgroupby([\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mIdentity-first\u001b[39m\u001b[38;5;124m'\u001b[39m])[\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mTitle\u001b[39m\u001b[38;5;124m'\u001b[39m]\u001b[38;5;241m.\u001b[39mnunique()\u001b[38;5;241m.\u001b[39msort_values(ascending\u001b[38;5;241m=\u001b[39m\u001b[38;5;28;01mFalse\u001b[39;00m)\u001b[38;5;241m.\u001b[39mhead(\u001b[38;5;241m10\u001b[39m))\n",
      "\u001b[1;31mAttributeError\u001b[0m: 'int' object has no attribute 'sort_values'"
     ]
    }
   ],
   "source": [
    "print(reviewed_matches['Person-first'].nunique().sort_values(ascending=False).head(10))\n",
    "print(\"\")\n",
    "print(reviewed_matches.groupby(['Identity-first'])['Title'].nunique().sort_values(ascending=False).head(10))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "id": "fa8328b3",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "PF_nouns\n",
      "              156\n",
      "patient        64\n",
      "child          26\n",
      "individual     24\n",
      "boy            10\n",
      "family          7\n",
      "subject         5\n",
      "proband         5\n",
      "case            4\n",
      "girl            4\n",
      "Name: Title, dtype: int64\n",
      "\n",
      "IF_nouns\n",
      "              134\n",
      "patient        82\n",
      "case           24\n",
      "child          18\n",
      "family         17\n",
      "individual     17\n",
      "group          10\n",
      "proband         9\n",
      "subject         8\n",
      "cohort          7\n",
      "Name: Title, dtype: int64\n"
     ]
    }
   ],
   "source": [
    "print(reviewed_matches.groupby(['PF_nouns'])['Title'].nunique().sort_values(ascending=False).head(10))\n",
    "print(\"\")\n",
    "print(reviewed_matches.groupby(['IF_nouns'])['Title'].nunique().sort_values(ascending=False).head(10))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 131,
   "id": "58342363",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "individual\n",
      "patients\n",
      "patient\n",
      "patient\n",
      "child\n",
      "subject\n",
      "child\n",
      "adolescent\n",
      "boy\n",
      "boy\n",
      "boy\n",
      "boy\n",
      "boy\n",
      "boy\n",
      "boy\n",
      "boy\n",
      "boy\n",
      "boy\n",
      "brother\n",
      "brother\n",
      "brother\n",
      "case\n",
      "case\n",
      "case\n",
      "case\n",
      "child\n",
      "child\n",
      "child\n",
      "child\n",
      "child\n",
      "child\n",
      "child\n",
      "child\n",
      "child\n",
      "child\n",
      "child\n",
      "child\n",
      "child\n",
      "child\n",
      "child\n",
      "child\n",
      "child\n",
      "child\n",
      "child\n",
      "child\n",
      "child\n",
      "child\n",
      "child\n",
      "child\n",
      "child\n",
      "child\n",
      "child\n",
      "child\n",
      "child\n",
      "child\n",
      "child\n",
      "child\n",
      "child\n",
      "child\n",
      "child\n",
      "child\n",
      "child\n",
      "cousin\n",
      "patient\n",
      "family\n",
      "family\n",
      "family\n",
      "family\n",
      "family\n",
      "family\n",
      "family\n",
      "girl\n",
      "girl\n",
      "girl\n",
      "girl\n",
      "individual\n",
      "individual\n",
      "individual\n",
      "individual\n",
      "individual\n",
      "individual\n",
      "individual\n",
      "individual\n",
      "individual\n",
      "individual\n",
      "individual\n",
      "individual\n",
      "individual\n",
      "individual\n",
      "individual\n",
      "individual\n",
      "individual\n",
      "individual\n",
      "individual\n",
      "individual\n",
      "individual\n",
      "individual\n",
      "individual\n",
      "individual\n",
      "individual\n",
      "individual\n",
      "male\n",
      "male\n",
      "patient\n",
      "patient\n",
      "patient\n",
      "patient\n",
      "patient\n",
      "patient\n",
      "patient\n",
      "patient\n",
      "patient\n",
      "patient\n",
      "patient\n",
      "patient\n",
      "patient\n",
      "patient\n",
      "patient\n",
      "patient\n",
      "patient\n",
      "patient\n",
      "patient\n",
      "patient\n",
      "patient\n",
      "patient\n",
      "patient\n",
      "patient\n",
      "patient\n",
      "patient\n",
      "patient\n",
      "patient\n",
      "patient\n",
      "patient\n",
      "patient\n",
      "patient\n",
      "patient\n",
      "patient\n",
      "patient\n",
      "patient\n",
      "patient\n",
      "patient\n",
      "patient\n",
      "patient\n",
      "patient\n",
      "patient\n",
      "patient\n",
      "patient\n",
      "patient\n",
      "patient\n",
      "patient\n",
      "patient\n",
      "patient\n",
      "patient\n",
      "patient\n",
      "patient\n",
      "patient\n",
      "patient\n",
      "patient\n",
      "patient\n",
      "patient\n",
      "patient\n",
      "patient\n",
      "patient\n",
      "patient\n",
      "patient\n",
      "patient\n",
      "patient\n",
      "patient\n",
      "patient\n",
      "patient\n",
      "patient\n",
      "patient\n",
      "patient\n",
      "patient\n",
      "patient\n",
      "patient\n",
      "patient\n",
      "patient\n",
      "patient\n",
      "patient\n",
      "patient\n",
      "patient\n",
      "patient\n",
      "patient\n",
      "patient\n",
      "patient\n",
      "patient\n",
      "patient\n",
      "patient\n",
      "patient\n",
      "patient\n",
      "patient\n",
      "patient\n",
      "patient\n",
      "patient\n",
      "patient\n",
      "patient\n",
      "people\n",
      "people\n",
      "people\n",
      "people\n",
      "people\n",
      "picture\n",
      "proband\n",
      "proband\n",
      "proband\n",
      "proband\n",
      "proband\n",
      "proband\n",
      "sibling\n",
      "subject\n",
      "subject\n",
      "subject\n",
      "subject\n",
      "subjects\n",
      "testing\n",
      "trio\n",
      "uncle\n",
      "youth\n"
     ]
    }
   ],
   "source": [
    "# np.array([reviewed_matches['PF_nouns']], [reviewed_matches['PF_nouns'].value_counts()])\n",
    "# If_test = np.array([reviewed_matches['IF_nouns']], [reviewed_matches['IF_nouns'].value_counts()])\n",
    "\n",
    "for thingy in reviewed_matches['PF_nouns']:\n",
    "    if thingy != \"\":\n",
    "        print(thingy)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 107,
   "id": "27a9a59a",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "PF_nouns\n",
      "              326\n",
      "              326\n",
      "              326\n",
      "individual     27\n",
      "              326\n",
      "             ... \n",
      "subjects        1\n",
      "testing         1\n",
      "trio            1\n",
      "uncle           1\n",
      "youth           1\n",
      "Name: Pf_counts, Length: 544, dtype: int64\n"
     ]
    }
   ],
   "source": [
    "Pf_test = pd.Series(reviewed_matches['PF_nouns'].value_counts(), index=reviewed_matches['PF_nouns'], name='Pf_counts')\n",
    "If_test = pd.Series(reviewed_matches['IF_nouns'].value_counts(), index=reviewed_matches['IF_nouns'], name='If_counts')\n",
    "\n",
    "print(Pf_test.)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1e8e4642",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9c040bb4",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a945d4bc",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7d2718e7",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "aeddb757",
   "metadata": {},
   "outputs": [],
   "source": [
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "id": "325e18cd",
   "metadata": {},
   "outputs": [
    {
     "ename": "NameError",
     "evalue": "name 'person_identity_count' is not defined",
     "output_type": "error",
     "traceback": [
      "\u001b[1;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[1;31mNameError\u001b[0m                                 Traceback (most recent call last)",
      "Cell \u001b[1;32mIn[39], line 1\u001b[0m\n\u001b[1;32m----> 1\u001b[0m \u001b[43mperson_identity_count\u001b[49m\u001b[38;5;241m.\u001b[39mplot()\n\u001b[0;32m      2\u001b[0m plt\u001b[38;5;241m.\u001b[39mshow()\n\u001b[0;32m      3\u001b[0m plt\u001b[38;5;241m.\u001b[39msavefig(\u001b[38;5;124m'\u001b[39m\u001b[38;5;124m..\u001b[39m\u001b[38;5;130;01m\\\\\u001b[39;00m\u001b[38;5;124moutput\u001b[39m\u001b[38;5;130;01m\\\\\u001b[39;00m\u001b[38;5;124mmatches_count.jpg\u001b[39m\u001b[38;5;124m'\u001b[39m)    \u001b[38;5;66;03m# we can right click on the plot above to save it, or save it via command\u001b[39;00m\n",
      "\u001b[1;31mNameError\u001b[0m: name 'person_identity_count' is not defined"
     ]
    }
   ],
   "source": [
    "person_identity_count.plot()\n",
    "plt.show()\n",
    "plt.savefig('..\\\\output\\\\matches_count.jpg')    # we can right click on the plot above to save it, or save it via command"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "f57fe19a",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "55\n",
      "94\n"
     ]
    }
   ],
   "source": [
    "person_examples = reviewed_matches.groupby(['Person-first'])['Person-first'].count()\n",
    "identity_examples = reviewed_matches.groupby(['Identity-first'])['Identity-first'].count()\n",
    "print(len(person_examples))\n",
    "print(len(identity_examples))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "ab1443d2",
   "metadata": {},
   "outputs": [],
   "source": [
    "person_identity_examples=pd.concat([person_examples,identity_examples],axis=1)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "11bf97e2",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Person-first</th>\n",
       "      <th>Identity-first</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>patient with autism</th>\n",
       "      <td>43.0</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>patient with ASD</th>\n",
       "      <td>41.0</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>child with ASD</th>\n",
       "      <td>16.0</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>child with autism</th>\n",
       "      <td>15.0</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>individual with autism</th>\n",
       "      <td>12.0</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>individual with ASD</th>\n",
       "      <td>10.0</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>proband with ASD</th>\n",
       "      <td>6.0</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>boy with autism</th>\n",
       "      <td>6.0</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>family with ASD</th>\n",
       "      <td>4.0</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>people with ASD</th>\n",
       "      <td>3.0</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                        Person-first  Identity-first\n",
       "patient with autism             43.0             NaN\n",
       "patient with ASD                41.0             NaN\n",
       "child with ASD                  16.0             NaN\n",
       "child with autism               15.0             NaN\n",
       "individual with autism          12.0             NaN\n",
       "individual with ASD             10.0             NaN\n",
       "proband with ASD                 6.0             NaN\n",
       "boy with autism                  6.0             NaN\n",
       "family with ASD                  4.0             NaN\n",
       "people with ASD                  3.0             NaN"
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "person_identity_examples.sort_values(by=['Person-first'], ascending=False).head(10)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "c04bc80a",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Person-first</th>\n",
       "      <th>Identity-first</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>ASD patient</th>\n",
       "      <td>NaN</td>\n",
       "      <td>77.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>autistic patient</th>\n",
       "      <td>NaN</td>\n",
       "      <td>22.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ASD case</th>\n",
       "      <td>NaN</td>\n",
       "      <td>22.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ASD family</th>\n",
       "      <td>NaN</td>\n",
       "      <td>19.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>autism patient</th>\n",
       "      <td>NaN</td>\n",
       "      <td>15.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>autistic child</th>\n",
       "      <td>NaN</td>\n",
       "      <td>14.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>autistic individual</th>\n",
       "      <td>NaN</td>\n",
       "      <td>12.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ASD subject</th>\n",
       "      <td>NaN</td>\n",
       "      <td>9.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ASD child</th>\n",
       "      <td>NaN</td>\n",
       "      <td>9.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ASD cohort</th>\n",
       "      <td>NaN</td>\n",
       "      <td>8.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                     Person-first  Identity-first\n",
       "ASD patient                   NaN            77.0\n",
       "autistic patient              NaN            22.0\n",
       "ASD case                      NaN            22.0\n",
       "ASD family                    NaN            19.0\n",
       "autism patient                NaN            15.0\n",
       "autistic child                NaN            14.0\n",
       "autistic individual           NaN            12.0\n",
       "ASD subject                   NaN             9.0\n",
       "ASD child                     NaN             9.0\n",
       "ASD cohort                    NaN             8.0"
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "person_identity_examples.sort_values(by=['Identity-first'], ascending=False).head(10)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "6e6184b4",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Person-first      55\n",
       "Identity-first    94\n",
       "dtype: int64"
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "person_identity_examples.notnull().sum()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6efde442",
   "metadata": {},
   "source": [
    "## Chart person-first or identity-first by year"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "47b82b88",
   "metadata": {},
   "source": [
    "## Count abstracts by the structures they use"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "be40279c",
   "metadata": {},
   "outputs": [],
   "source": [
    "test = pd.merge(person_by_title,identity_by_title,on='Title',how='outer')\n",
    "print(test)\n",
    "print(type(test))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7dd85bf7",
   "metadata": {},
   "outputs": [],
   "source": [
    "person_by_title = reviewed_matches.groupby(['Title'])['Person-first'].count()\n",
    "identity_by_title = reviewed_matches.groupby(['Title'])['Identity-first'].count()\n",
    "title = pd.concat([person_by_title,identity_by_title],axis=1)\n",
    "print(title)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "975b4d8a",
   "metadata": {},
   "outputs": [],
   "source": [
    "title.sort_values(by=['Identity-first'], ascending=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f5f461ab",
   "metadata": {},
   "outputs": [],
   "source": [
    "title.sort_values(by=['Person-first'], ascending=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4df4af17",
   "metadata": {},
   "outputs": [],
   "source": [
    "columns = ['Person-first','Identity-first']\n",
    "filter_ = (title[columns] > 0).all(axis=1)\n",
    "title[filter_]\n",
    "len(title[filter_])\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "025e8bad",
   "metadata": {},
   "outputs": [],
   "source": [
    "title[filter_].sort_values(by=['Person-first'], ascending=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "13737d2d",
   "metadata": {},
   "outputs": [],
   "source": [
    "title[filter_].sort_values(by=['Identity-first'], ascending=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7906c98e",
   "metadata": {},
   "outputs": [],
   "source": [
    "has_pf = title[title['Person-first'] > 0]\n",
    "has_both = has_pf[has_pf['Identity-first'] > 0]\n",
    "print(len(has_both))\n",
    "has_both"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "43c3632b",
   "metadata": {},
   "outputs": [],
   "source": [
    "plt.scatter(title['Person-first'], title['Identity-first'])\n",
    "plt.show()\n",
    "\n",
    "#plt.savefig('..\\\\output\\title_count.jpg')    # we can right click on the plot above to save it, or save it via command"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d07e4e72",
   "metadata": {},
   "outputs": [],
   "source": [
    "title['Person-first']\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b8ab14d1",
   "metadata": {},
   "outputs": [],
   "source": [
    "ax = plt.axes()\n",
    "ax.scatter(x, y, c='g', marker='x')\n",
    "ax.set_title(\"Anscombe's First Data Set\")\n",
    "ax.set_ylabel('Y-Values')\n",
    "ax.set_ylim(4, 11)\n",
    "ax.set_xlabel('X-Values')\n",
    "ax.set_xlim(3, 15)\n",
    "\n",
    "plt.show()"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.1"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
